text/0017-parameter_type_with_encoding.md

```diff
@@ -0,0 +1,143 @@
+  A 32-bit integer is sent as 32 bits of data.
+- **C-style cast:** The parameter value is converted to a `float`.
+  This may result in some loss of precision as a `float` can represent an integer with up to 24 bits of pecision.
+
+A component should support only one type, and indicate that type by setting the `MAV_PROTOCOL_CAPABILITY_PARAM_ENCODE_BYTEWISE` (byte-wise encoding) or `MAV_PROTOCOL_CAPABILITY_PARAM_ENCODE_C_CAST` (C-style encoding) protocol bits in `AUTOPILOT_VERSION.capabilities`.
+A GCS may support both types and use the type that is indicated by the target component.
+
+Byte wise encoding is recommended as it allows very large integers to be exchanged (e.g. 1E7 scaled integers can be useful for encoding some types of data, but lose precision if cast directly to floats).
+
+The above approach allows a GCS to determine the parameter type and encoding, and from that extract the value.
+
+This RFC proposes a parallel approach where both the parameter data type _and_ its encoding are represented in the `param_type` field.
+This allows a GCS (or other system) to be able to decode a particular parameter type without access to any other information. 
+This in turn means that parameters can be decoded from a partial log. 
+What it also means is that it provides a migration path for an autopilot from one encoding to the other, because the switch does not have to be done in one release and can be staggered as needed.
+
+
+# Detailed Design
+
+The design has two parts:
+- New enum values in [MAV_PARAM_TYPE](https://mavlink.io/en/messages/common.html#MAV_PARAM_TYPE) that include the parameter type and encoding.
+  These allow the GCS to decode each param based on its type.
+- [`PARAM_REQUEST_READ`](https://mavlink.io/en/messages/common.html#PARAM_REQUEST_READ) and [`PARAM_REQUEST_LIST`](https://mavlink.io/en/messages/common.html#PARAM_REQUEST_LIST) are extended with an 32 bit bitmask indicating which which `MAV_PARAM_TYPE` flags that the GCS/requester understands.
+  The flight stack would cache this response and use the appropriate type for each parameter.
+  This allows the flight stack to send only types supported by the corresponding GCS.
+
+
+## New Enum types
+
+We would add new [MAV_PARAM_TYPE](https://mavlink.io/en/messages/common.html#MAV_PARAM_TYPE) values corresponding to each of the existing types, but including 
+the encoding.
+A recipient can therefore determine the type and encoding from the message, without having to query any other source.
+
+The proposed pattern for the 10 new "cast" and 10 new "bytewise" flags is: `MAV_PARAM_TYPE_[BYTEWISE|CAST]_[UINT8|INT8|UINT16|INT16|UINT32|INT32|UINT64|INT64|REAL32|REAL64]`.
+So for example, to send 32-bit integers you could use `MAV_PARAM_TYPE_INT32`, `MAV_PARAM_TYPE_BYTEWISE_INT32`, or `MAV_PARAM_TYPE_CAST_INT32` (by preference you would send using `MAV_PARAM_TYPE_BYTEWISE_INT32` if this was understood by the GCS.
+
+Question? 
```

Comment:
@tridge a few questions here inline - e.g. do we need to indicate support for the old bits? Do we need flags for c-style encoding - after all, wouldn't you update in one go to the new type?